File: 002-charclass-unicode/matrix.mjs
#!/usr/bin/env node
/**
 * 同じ日本語入力を JavaScript の正規表現にかけて、Python との差を見る。
 *
 * Python 側（matrix.py）と並べて読むと、文字クラスの判定そのものより
 * 「1 文字をどう数えるか」の違いが効いてくることが分かる。
 */

const SUBJECTS = [
  ['半角数字', '0123'],
  ['全角数字', '０１２３'],
  ['丸数字', '①②③'],
  ['半角英字', 'abc'],
  ['全角英字', 'ａｂｃ'],
  ['漢字', '日本語'],
  ['半角カナ', 'ｱｲｳ'],
  ['絵文字（BMP 外）', '🍣🍺'],
  ['結合文字（が = か + 濁点）', 'が'],
];

const CLASSES = [
  ['\\d', (s) => /^(?:\d)+$/.test(s)],
  ['[0-9]', (s) => /^(?:[0-9])+$/.test(s)],
  ['\\w', (s) => /^(?:\w)+$/.test(s)],
  ['.', (s) => /^(?:.)+$/.test(s)],
];

const mark = (ok) => (ok ? '○' : '×');

console.log(`Node ${process.version}\n`);

let header = '入力'.padEnd(30) + '例'.padEnd(12);
for (const [label] of CLASSES) header += label.padEnd(8);
header += 'length  実文字数';
console.log(header);
console.log('-'.repeat(header.length));

for (const [label, text] of SUBJECTS) {
  let line = label.padEnd(30) + text.padEnd(12);
  for (const [, test] of CLASSES) line += mark(test(text)).padEnd(8);
  // length は UTF-16 の単位数、[...text] は符号位置の数
  line += String(text.length).padEnd(8) + String([...text].length);
  console.log(line);
}

console.log('\n■ ドットは絵文字を 1 文字として扱えるか');
const emoji = '🍣';
console.log(`  /^.$/ が ${emoji} に一致するか        : ${mark(/^.$/.test(emoji))}  ← u フラグなし`);
console.log(`  /^.$/u が ${emoji} に一致するか       : ${mark(/^.$/u.test(emoji))}  ← u フラグあり`);
console.log(`  "${emoji}".length                     : ${emoji.length}（UTF-16 の単位数）`);
console.log(`  [..."${emoji}"].length               : ${[...emoji].length}（符号位置の数）`);

console.log('\n■ Unicode プロパティ（Python の標準 re には無い）');
console.log(`  /^\\p{Script=Han}+$/u が 日本語 に一致 : ${mark(/^\p{Script=Han}+$/u.test('日本語'))}`);
console.log(`  /^\\p{Nd}+$/u が ０１２３ に一致       : ${mark(/^\p{Nd}+$/u.test('０１２３'))}`);

console.log('\n■ v フラグ（ES2024）の集合演算');
try {
  const subtraction = new RegExp('^[\\p{ASCII}--[0-9]]+$', 'v');
  console.log(`  集合差 [\\p{ASCII}--[0-9]] が abc に一致 : ${mark(subtraction.test('abc'))}`);
  console.log(`  同じものが 123 に一致                   : ${mark(subtraction.test('123'))}  ← 数字は引かれている`);
  const intersection = new RegExp('^[\\p{Script=Han}&&\\p{Letter}]+$', 'v');
  console.log(`  積集合 [\\p{Script=Han}&&\\p{Letter}] が 日本語 に一致 : ${mark(intersection.test('日本語'))}`);
} catch (e) {
  console.log(`  v フラグが使えません: ${e.message}`);
}

console.log('\n■ 覚えておくこと');
console.log('  ・\\d は JavaScript では半角数字だけです（Python は全角にも一致します）');
console.log('  ・u フラグなしのドットは絵文字を 2 つに割ります');
console.log('  ・結合文字は見た目が 1 文字でも 2 つの符号位置です');
console.log('  ・文字クラスの引き算・積は v フラグ（ES2024・Node 20.12 以降）で書けます');

File: 002-charclass-unicode/matrix.py
#!/usr/bin/env python3
"""日本語の入力が文字クラスでどう扱われるかを一覧にする。

「電話番号の欄に全角で入力されたら弾かれた」という事故は、
文字クラスが何を数字とみなすかを確かめていないところから起きる。

ここでは同じ入力を複数の文字クラスにかけ、さらに Unicode 正規化を
通した場合と通さない場合を並べて出す。
"""

import re
import sys
import unicodedata

SUBJECTS = [
    ("半角数字", "0123"),
    ("全角数字", "０１２３"),
    ("丸数字", "①②③"),
    ("半角英字", "abc"),
    ("全角英字", "ａｂｃ"),
    ("漢字", "日本語"),
    ("半角カナ", "ｱｲｳ"),
    ("絵文字（BMP 外）", "🍣🍺"),
    ("結合文字（が = か + 濁点）", "が"),
]

CLASSES = [
    (r"\d", "数字とみなすか"),
    (r"[0-9]", "半角数字に限るか"),
    (r"\w", "単語構成文字とみなすか"),
    (r".", "1 文字として数えるか"),
]


def probe(pattern: str, text: str) -> str:
    """text 全体がそのクラスの繰り返しで表せるかを見る。"""
    return "○" if re.fullmatch(f"(?:{pattern})+", text) else "×"


def main() -> int:
    print(f"Python {sys.version.split()[0]}\n")

    header = f"{'入力':<26}{'例':<10}" + "".join(f"{p:<8}" for p, _ in CLASSES) + "文字数"
    print(header)
    print("-" * len(header))
    for label, text in SUBJECTS:
        cells = "".join(f"{probe(p, text):<8}" for p, _ in CLASSES)
        print(f"{label:<26}{text:<10}{cells}{len(text)}")

    print("\n■ Unicode 正規化を通すと何が変わるか（NFKC）")
    print(f"{'入力':<26}{'元':<10}{'NFKC 後':<12}{'\\d に一致':<12}")
    print("-" * 62)
    for label, text in SUBJECTS:
        normalized = unicodedata.normalize("NFKC", text)
        before = probe(r"\d", text)
        after = probe(r"\d", normalized)
        changed = "→ 変化あり" if before != after else ""
        print(f"{label:<26}{text:<10}{normalized:<12}{before} → {after}  {changed}")

    print("\n■ 覚えておくこと")
    print("  ・\\d は Python では全角数字にも一致します（[0-9] は一致しません）")
    print("  ・NFC ではなく NFKC を通すと、全角は半角へ、丸数字は数字へ変換されます")
    print("  ・絵文字は Python では 1 文字ですが、JavaScript では 2 単位として扱われます")
    print("    （matrix.mjs で同じ入力を試すと差が見えます）")
    return 0


if __name__ == "__main__":
    sys.exit(main())
